add back copy stageout plugin

hufnagel opened the pull request:
Fixes https://github.com/dmwm/WMCore/issues/9144
Not just the same old Backends/CPImpl, but a rewrite. Added stagein support since this exists in XRDCPImpl, which this new version is based on. Do we even still support stagein? If not maybe we should clean that up, it makes the code much more complicated than it has to be.

The Plugins/CPImpl is similar to the old version, with a few minor tweaks/changes.

@amaltaro , what do you think?

hufnagel committed:
add back copy stageout plugin, fixes #9144

Add back CPImpl, a basic cp based stageout plugin. Needed for resources with large shared file systems like HPC.

hufnagel commented:
Completely untested, just wanted to get some early feedback on code structure etc. Also not sure how I would test this.

amaltaro commented:
Dirk, I think we can also copy back those 3 test changes. BTW, you have pushed in an empty CPImpl at the root directory as well.

hufnagel commented:
Added back CPImpl unit test as-is and also adjusted the other unit test. Seems to me we might want to get rid of the special test-copy unit test implementation altogether since it isn't really used now anymore.

hufnagel commented:
@amaltaro , how can I get cmsdmwmbot to actually give me the output of the unit test?

amaltaro commented:
Given the list of failures in the report.html, you look them up here:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/8732/#showFailuresLink

which gives you a link to, for instance:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/8732/testReport/junit/WMCore_t.Storage_t.Backends_t.CPImpl_t/CPImplTest/testCreateOutputDirectory_simple/

I'm not sure how up-to-date this is, but it might be helpful as well:
https://github.com/dmwm/WMCore/wiki/Understanding-Jenkins

hufnagel commented:
Thank, that helps, I see what's wrong now. Would be good if the report that is linked in the standard reply would contain that link...

hufnagel commented:
Ok, this looks good now. @amaltaro , can you review?

Also, what about the other uses of "test-copy"? Should we clean this up?

amaltaro commented:
The link is made available via "default" check, just push the "Details" link.
If "test-copy" is no longer used anywhere, then I think we can remove it (but I have the impression there are still a few tests using it).

amaltaro commented:
BTW, you can ask for reviewers at the top right button "Reviewers". It's easier than to find what is waiting for your review.

hufnagel commented:
I don't think test-copy is actually used anywhere now. There are a few references left though:

[hufnagel@lxplus717 WMCore]$ grep -r test-copy test
test/python/WMCore_t/WMRuntime_t/Scripts_t/SetupCMSSWPset_t.py:            <command value="test-copy"/>
test/python/WMCore_t/WMSpec_t/samples/site-local-config.xml:    <command value="test-copy"/>
test/python/WMCore_t/Misc_t/site-local-config.xml:    <command value="test-copy"/>
[hufnagel@lxplus717 WMCore]$ grep -r test-copy src
src/python/WMCore/Storage/Backends/UnittestImpl.py:registerStageOutImpl("test-copy", LocalCopyImpl)
src/python/WMCore/Storage/Registry.py:                'test-copy': 'TestLocalCopyImpl',

hufnagel commented:
Ok, requested a review. Even if we want to get rid of these last "test-copy" reference (which at first glance don't seem to be functional to me), that should happen in a separate PR.

amaltaro commented on src/python/WMCore/Storage/Plugins/CPImpl.py:
```diff
@@ -0,0 +1,43 @@
+
+    def doTransfer(self, fromPfn, toPfn, stageOut, pnn, command, options, protocol, checksum ):
+        self.createOutputDirectory(toPfn)
+        shutil.copy(fromPfn, toPfn)
```
This `shutil.copy()` doesn't copy all the file metadata. But looking at the history, that's how this plugin used to be, so I think we can leave it like that.

amaltaro reviewed (approved):
Patch looks good, Dirk. Is there any place we can run a few jobs to get it tested before we merge it?

In addition to that, I have two comments:
* when describing your changes in the PR, you can write "Fixes #9144" such that it also closes the issue when it gets merged
* please keep test/* changes and pylint/pep fixes in a different commit than the one with the real src/* changes. No need to change this one, but keep it in mind for future ones please.

hufnagel commented:
There is no site that uses this as a stageout plugin, that's why I said testing this right now is hard.

amaltaro commented:
Ok, we shall eventually see whether it works or not. Thanks
